staging/src/k8s.io/apiextensions-apiserver/pkg/registry/customresourcedefinition/etcd.go

```diff
@@ -127,6 +127,9 @@ func (r *REST) Delete(ctx context.Context, name string, deleteValidation rest.Va
 				if existingCRD.DeletionTimestamp.IsZero() {
 					now := metav1.Now()
 					existingCRD.DeletionTimestamp = &now
+					// Setting the grace period to 0, so that once the finalizers are removed from the crd, apiserver deletes the crd.
+					var zero int64 = 0
+					existingCRD.DeletionGracePeriodSeconds = &zero
```

Comment:
this shouldn't be required, since CRD types don't implement graceful deletion, right?

Reply:
If the GracePeriod isn't set to 0, then after an UPDATE that removes all the finalizers, the apiserver will not automatically delete-on-update. See this line in `shouldDeleteDuringUpdate`:

https://github.com/kubernetes/kubernetes/blob/36e06bbb1fd3fc430ff4f8990684bbd53cf80e79/staging/src/k8s.io/apiserver/pkg/registry/generic/registry/store.go#L405

Reading this again, maybe we should fix `shouldDeleteDuringUpdate` instead, the condition should be 
```golang
func shouldDeleteDuringUpdate() bool {
  noGracePeriod := oldMeta.GetDeletionGracePeriodSeconds() == nil || *oldMeta.GetDeletionGracePeriodSeconds() == 0
  return len(newMeta.GetFinalizers()) == 0 && oldMeta.DeletionTimestamp() != nil && noGracePeriod
}
```